src/fontra/views/editor/edit-tools-shape.js

```diff
@@ -0,0 +1,202 @@
+
+export class ShapeToolRect extends BaseTool {
+  iconPath = "/tabler-icons/square-plus-2.svg";
+  identifier = "shape-tool-rectangle";
+
+  setCursor() {
+    if (this.sceneModel.selectedGlyph?.isEditing) {
+      this.canvasController.canvas.style.cursor = "crosshair";
+    }
+  }
+
+  handleHover(event) {
+    if (!this.sceneModel.selectedGlyph?.isEditing) {
+      this.editor.tools["pointer-tool"].handleHover(event);
+      return;
+    }
+    this.setCursor();
+  }
+
+  async handleDrag(eventStream, initialEvent) {
+    if (!this.sceneModel.selectedGlyph?.isEditing) {
+      await this.editor.tools["pointer-tool"].handleDrag(eventStream, initialEvent);
+      return;
+    }
+
+    const initialPoint = this.sceneController.selectedGlyphPoint(initialEvent);
+
+    if (!(await shouldInitiateDrag(eventStream, initialEvent))) {
+      // TODO: open dialog for numeric size input
+      return;
+    }
+
+    let mouseRect;
+    let eventTemp;
+    for await (const event of eventStream) {
+      eventTemp = event;
+      const point = this.sceneController.selectedGlyphPoint(event);
+      if (point.x === undefined) {
+        // We can receive non-pointer events like keyboard events: ignore
+        continue;
+      }
+      mouseRect = rectangle.rectRound({
+        xMin: initialPoint.x,
+        yMin: initialPoint.y,
+        xMax: point.x,
+        yMax: point.y,
+      });
+      const drawPath = new Path2D();
+      this.drawShapePath2D(drawPath, mouseRect, event);
+      this.sceneModel.shapeToolShapePath = drawPath;
+      this.sceneModel.event = event;
+      this.canvasController.requestUpdate();
+    }
+
+    delete this.sceneModel.shapeToolShapePath;
+    delete this.sceneModel.event;
+    this.canvasController.requestUpdate();
+
+    // rectsize return when too small
+    if (!mouseRect) {
+      return;
+    }
+
+    const pathNew = this.drawShapeVarPackedPath(mouseRect, eventTemp);
+    // reversed contour direction
+    if (eventTemp.ctrlKey) {
+      this.reversePath(pathNew);
+    }
+    this.addShapePath(pathNew);
+  }
+
+  async addShapePath(pathNew) {
+    await this.sceneController.editGlyphAndRecordChanges(
+      (glyph) => {
+        const editLayerGlyphs = this.sceneController.getEditingLayerFromGlyphLayers(
+          glyph.layers
+        );
+
+        const firstLayerGlyph = Object.values(editLayerGlyphs)[0];
+        const selection = new Set();
+        const firstIndex = firstLayerGlyph.path.numPoints;
+        for (const index of range(pathNew.numPoints)) {
+          const point = pathNew.getPoint(index);
+          if (!point.type) {
+            selection.add(`point/${firstIndex + index}`);
+          }
+        }
+        this.sceneController.selection = selection;
+
+        for (const [layerName, layerGlyph] of Object.entries(editLayerGlyphs)) {
+          layerGlyph.path.appendPath(pathNew);
+        }
+        return "add shape";
```

Comment:
Would be nice if this could be "rectangle", "square", "ellipse" or "circle".

Suggest to add a static attr to each shape class:
   `static shapeNames = ["rectangle", "ellipse"];`
And index by shift key:
```js
    return `add ${this.constructor.shapeNames[shiftKey ? 1 : 0]}`
```

Reply:
I am not 100% sure what you want. The following change will at least be a bit more precise and still be flexible if we would extend the tools:
https://github.com/googlefonts/fontra/blob/2426d9a003298f51295fb5596a94a05994c125aa/src/fontra/views/editor/edit-tools-shape.js#L98

Reply:
I reverted this change, because it caused a fail in the tests.

Reply:
(Sometimes the node tests glitch. The failure was not caused by this change.)

Sorry, my original comment was sloppy. I meant this (and forget about `static` and `.constructor`):

Rectangle tool:
```js
  shapeNames = ["rectangle", "square"];
```
Ellipse tool:
```js
  shapeNames = ["ellipse", "circle"];
```

And for the undo label:
```js
return `add ${this.shapeNames[shiftKey ? 1 : 0]}`
```

Reply:
OK, thanks. Done.